ET-4971 doc for split documents (#1078)
xaynetwork/xayn_discovery_engine: docs/source/getting_started/getting_started.md

Here we take a high-level look at how the system works. The API is divided into two parts: the back office and the front office.

The back office system can be used to ingest your content into the system. Every piece of content you ingest is stored as one document with one or more snippets.

> A snippet is a bit of text which the user might search for. In the simplest case you only have one snippet per document in which case the snippet is roughly equivalent to the document. If you have such a use case you can mentally replace snippet with document in most of the remaining documentation. Through as there are limits on

Reviewer: > A snippet is a bit of text which the user might search for.
I think it's more clear to present snippets as pieces or the whole content of a document.

```suggestion
A document is comprised of one or more snippets. Each snippet represents a piece of the content of the document. In the simplest case, you only have one snippet per document in such a case you can mentally replace snippet with document in most of the remaining documentation. Through as there are limits on
```

this can be used to get better result on larger documents. More information on this can be found in the section  [split documents](#split-documents).

The `properties` field is completely optional. It can contain custom data that can be used for filtering and that the system will return when a document is part of the result of a query.

The data that can be included in the properties is limited in terms of type and size. We support numbers, strings, boolean, date and list of strings, none of which are nullable. Please see [createDocuments](https://docs.xayn.com/back_office.html#tag/documents/operation/createDocuments) for more information on properties.

This example assumes that we will eventually display the returned documents as a 'for-you' section, where we want to display an article's image, title, text preview, and a link (for click-through), so we have included these specific properties during ingestion.

## Split Documents

We provide a functionality to extract multiple snippets from the provided content of a document.

The system uses Natural language processing (NLP) algorithms to split the document into multiple parts.

> 

Author: :pushpin: https://github.com/xaynetwork/xayn_discovery_engine/pull/1080

placed it into follow up PR

```

```{note}

> Please note that if an interaction between a user and a document is added, all snippets of a document will **not** be part of the documents returned for future calls to the personalised endpoint.

Author: :pushpin:  When updating this line I realized that semantic search is currently quite hard to use for anything which has documents with multiple snippets (except for a use case of refining existing search).

The problem is if you don't do semantic search based on previous search results you have no idea which snippet ids you want to use as the customer only knows the document ids and that there might be more then one snippet but not how many and which snippets there are :thinking:

> Please note that if an interaction between a user and a document is added, all snippets of a document will **not** be part of the documents returned for future calls to the personalised endpoint.

Reviewer: I would keep the document in focus and then specify about the snippets.

```suggestion
Please note that if an interaction between a user and a document is added, the document will **not** be part of the documents returned for future calls to the personalised endpoint. This includes all snippets associated to that document```

{
  "documents": [
  {

>       "snippet_id": { "document_id": "xayn_5283ef3", "sub_id": 0 },

Reviewer: Does this happen even if you never use the split option? I think that for backward compatibility if split is never used this example should stay the same.
If this is the case would be better to specify this and show it with two different examples.

Reviewer: Ok, is specified underneath, but maybe we can find a way to specify both? 
In the ingestion example we never used "split = true".

Author: The `snippet_id` is  always returned the system doesn't know (or care) if `split` was used or not.

For backward compatibility we still return the previous `id` but it's deprecated in the OpenApi spec, so I didn't include it in the example.

Reviewer: ok thanks.

In the request, we ask the system to include the properties of the returned documents. We can use this data to implement a 'more like this' section.

We also have a `score` field which represents how well the documents match the user's interests. The higher the number, the better the documents match. It should be noted that the scores only have meaning in relation to other
scores from the same requests.

> `snippet_id` is the id of the snippet which was found. The `document_id` in the `snippet_id` is the id of the document associated with the snippet. If you do not use ingestion options like `split` and in turn only have

Reviewer: ```suggestion
The field `snippet_id` identifies a specific snippet. The `document_id` in the `snippet_id` is the id of the document associated with the snippet. If you do not use ingestion options like `split` and in turn only have
```

Reviewer: I would make `If you do not use ingestion options like split` the start of a new paragraph. This part is very important and it is easier that will be noted.

one snippet per document then you can always use `snippet_id.document_id` and ignore the rest. In many places
both a the full snippet id object or a document id string can be used. If documents which have multiple snippets are ingested it's highly recommended to always use the full snippet id.

# Search

Depending on the use-case searching for documents can be achieved as a search for documents _similar_ to a given snippet/document or as a _free-text search_. Both variants can then be run as a anonymous search or a search that is personalized. Personalization comes in two fashions, with a _user-id_ or by providing a interaction _history_.

## Similar documents

> In this search variant either a _document id_ string or a _snippet id_ object must be provided to the [`/semantic_search`](https://docs.xayn.com/front_office.html#tag/front-office/operation/getSimilarDocuments) endpoint.

Reviewer: We already showed what a document and snippet ids are. I wouldn´t specify that they are string and objects.
```suggestion
In this search variant either a _document id_ or a _snippet id_ must be provided to the [`/semantic_search`](https://docs.xayn.com/front_office.html#tag/front-office/operation/getSimilarDocuments) endpoint.
```

```bash
curl -X POST "$URL/semantic_search" \
    --header "authorizationToken: $FRONTOFFICE_TOKEN" \
    --header "Content-Type: application/json" \
    --data '{
        "document": { "id": { "document_id": "xayn_cd5604c", "sub_id": 2 } },
        "include_properties": true
    }'
```

> The result contains a list of snippets that are similar to the identified snippet/document.

Reviewer: We pass a snippet id
```suggestion
The result contains a list of snippets that are similar to the identified snippet.
```

If only documents with a single snippet are used you can provide only the document id
like this: `"document": { "id": "xayn_cd5604c" },`.

> Be aware that until our implementation is further improved this is equivalent to doing
> a semantic search on the first snippet of a document.

Reviewer: ```suggestion
Be aware that at this moment, this is equivalent to doing a semantic search on the first snippet of a document.
In the future, this behaviour could change to better represent the intent to search for something similar to the whole document. If you want to explicitly search for the first snippet use a _snippet id_.
```

```

```{note}
Please note, that setting candidates can only be undone by sending the complete list of all ingested document-ids again.
```

The candidates can facilitate fast transitions between different sets of documents without compromising the users' centers of interest (COIs) with which they were engaging. One practical scenario is handling outdated news articles that should not reappear in the recommendations. However, the past user interactions with those outdated articles should still influence the suggested documents.

> # Split Documents

Reviewer: I would move this to a subsection of the ingestion at the beginning.

We provide a functionality to extract multiple snippets from the provided content of a document.

> This is roughly equivalent to splitting the document into multiple parts.
> 
> We try to have natural language boundaries between snippets as far as possible. For example while the
> algorithm tries to split at sentence boundaries it doesn't do so by naively splitting based on dots as
> they are also e.g. used with abbreviations. Instead it uses a natural language aware approach.

Reviewer: Too many implementation details.

```suggestion
The system uses Natural language processing (NLP) algorithms to split the document into multiple parts.
```

> Currently the splitting will only work reliable well with text of the language configured by Xayn for
> a specific tenant (i.e. a specific front office/back office instance pair). We are currently working on integrating language detection to make this more convenient usable and allow having documents of different
> languages in the same system.

Reviewer: ```suggestion
Currently, automatic splitting works only with one language set when the system is set up; by default, it is English. If you need another one, please contact us.
We are working to add support for multiple languages to our text-splitting algorithms.
```
